Add .NET SDK for Terraform provider

mikhailshilkov opened the pull request:
Resolves #516

mikhailshilkov commented on sdk/dotnet/State/ArtifactoryRemoteStateReferenceArgs.cs:
```diff
@@ -0,0 +1,59 @@
+    /// <summary>
+    /// The configuration options for a Terraform Remote State stored in the Artifactory backend. 
+    /// </summary>
+    public class ArtifactoryRemoteStateReferenceArgs : RemoteStateReferenceArgs
```
@pgavlin See https://github.com/pulumi/pulumi-terraform/pull/517#discussion_r368624262

pgavlin replied:
This definitely is not something that the schema supports. Do we need to use an approach like this?

mikhailshilkov replied:
How can we model sum types? How is the [TypeScript version](https://github.com/pulumi/pulumi-terraform/blob/master/sdk/nodejs/state/remoteStateReference.ts#L33) represented in the schema?

pgavlin replied:
We can't at the moment. FWIW, this is the only SDK that uses this pattern.

mikhailshilkov replied:
Should we remodel as N resources with N independent args types?

pgavlin replied:
That would work. We could also use one resource with one arg type that supports all of the properties needed for the various backends, then use overlays for a more ergonomic experience.

mikhailshilkov replied:
I think separate resources make for a better user experience. I changed the current implementation to the N resources model.

pgavlin replied:
We should also change the type tokens s.t. they reflect the resource type. Note that this will require changing the provider implementation accordingly.

mikhailshilkov replied:
That sounds like a breaking change and 2.0 then?
Do they have to match?
Should we proceed with the original implementation for now and leave this for when the resource model is there, if it's breaking anyway?

mikhailshilkov replied:
Besides, we'd need to fix master vs dotnet branches before that...

pgavlin replied:
We don't have to make that a breaking change: we can continue to accept the original type token as well as the new, more specific tokens.

We could also treat the union args as overlays and let the schema describe the union of the args type. If that's not clear, lmk and I'll put together an example.

mikhailshilkov replied:
Overlays mean hand-written specific types that are mapped to one internal union type?

pgavlin replied:
That's right.

mikhailshilkov replied:
Wouldn't that be exactly equivalent to my current PR in terms of public API and resource type token?

pgavlin replied:
It'd be very close, yeah. There would be one additional resource + args type for the single actual resource (i.e. `RemoteStateReference`).

mikhailshilkov replied:
Would `RemoteStateReference` be the base class then (but not the args)?

pgavlin replied:
It could be. I don't think it has to be, but if that makes the most sense for the SDK then that's probably what we should do.

mikhailshilkov replied:
I gave it a thought and it feels very awkward to me. A base class `RemoteStateReference` class that nobody should actually use (they should use overlays instead) and monster `RemoteStateReferenceArgs` which includes all properties from all args like `IamEndpoint` and `ResourceGroupName`...
@CyrusNajmabadi could you join the discussion above and see what you think?

CyrusNajmabadi commented on examples/localstate-dotnet/.gitignore:
```diff
@@ -0,0 +1,353 @@
+MigrationBackup/
+
+# Ionide (cross platform F# VS Code tools) working folder
+.ionide/
```
is thsi entire .gitignore desirable here? is it a copy of something somewhere else?

mikhailshilkov replied:
I believe this is from a github repo of standard gitignores, the one for .NET

CyrusNajmabadi commented on sdk/dotnet/State/ArtifactoryRemoteStateReference.cs:
```diff
@@ -0,0 +1,98 @@
+// Copyright 2016-2020, Pulumi Corporation.
```
how was this file generated?

mikhailshilkov replied:
This whole PR is manually written

jen20 replied:
There is currently no way to generate these since the Terraform provider is not bridged. The same approach was used for TypeScript and Python. Fortunately they do not change very often, however!

mikhailshilkov commented:
The summary of the offline discussion:
1. Terraform has a single resource, that's why we implemented a single resource with a single type tag in the provider.
2. The initial goal was to use tfbridge for the SDKs but it proved to be hard because this resource is implemented in core Terraform.
3. We then proceeded with the manual creation of the node/python SDKs.
4. While doing so, we defined one resource with the argument type which is a discriminated union of all possible argument types, discriminated by the `backend` property.
5. There's no easy way to represent such discriminated unions in the current tfgen's schema.
6. In the absence of a real decision about whether we actually need to support this API in the schema, we are going forward with the approach we have today also for the .NET SDK.

@pgavlin @jen20 @lukehoban I reverted the PR to have one resource with N argument classes derived from a common base class, which is a C# approximation of discriminated unions.

mikhailshilkov commented:
Closing in favor of #526
